static-analysis/teammates-macker.xml

```diff
@@ -1,6 +1,161 @@
+                </allow>          
+            </deny>
+        </access-rule>
+    </ruleset>
+    
+    <ruleset name="UiComponent: Only *Action in can touch logic api">
+        <access-rule>
+            <deny>
+                <from class="teammates.ui.**" />
+                <to class="teammates.logic.**" />   
+                <allow>
+                    <from>
+                        <include class="teammates.ui.controller.**" />
+                        <include class="teammates.ui.automated.**" />
+                    </from>
+                    <to class="teammates.logic.api.**" />
+                </allow>
+                <allow>
+                    <!-- to import sample data, need to be changed -->
+                    <from class="teammates.ui.controller.AdminInstructorAccountAddAction" />
+                    <to class="teammates.logic.backdoor.BackDoorLogic" />
+                </allow>    
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="UiComponent: *PageData is only used in *Action and ActionResult">
+        <access-rule>
+            <deny>
+                <to class="teammates.ui.teamplate.**" />
+                <allow>
+                    <from>
+                        <include class="teammates.ui.controller.**" />
+                        <include class="teammates.ui.template.**" />
+                    </from>
+                </allow>          
+            </deny>
+        </access-rule>
+    </ruleset>
+    
+    <ruleset name="LogicComponent: Each logic can only access its corresponding DB (e.g. AccountsLogic -> AccountsDb)">
+        <foreach var="logicModule" class="teammates.logic.core.(*)Logic">
+            <pattern name="DbModule" class="teammates.storage.api.${logicModule}Db" />
+            
+            <access-rule>  
+                <message>${logicModule} can only access its corresponding DB</message>
+                <deny> 
+                    <from class="teammates.logic.core.${logicModule}Logic" />
+                    <to class="teammates.storage.api.**" /> 
+                    <allow>
+                        <to pattern="DbModule"></to>
+                    </allow>
+                </deny>
+            </access-rule>
+        </foreach>
+    </ruleset>
+    <ruleset name="LogicComponent: BackDoorServlet should only access BackDoorLogic, not logic:core">
+        <access-rule>
+            <deny>
+                <from class="teammates.logic.backdoor.BackDoorServlet" />
+                <to class="teammates.logic.core.**" />      
+            </deny>
+        </access-rule>
+    </ruleset>
+    
+    <ruleset name="StorageComponent: storage::entity should only be accessed by storage::api and common:datatransfer">
+       <access-rule>
+            <deny>
+                <to class="teammates.storage.entity.**" />
+            </deny>
+            <allow>
+                <from>
+                    <include class="teammates.test.**" />
```

Comment:
I don't think this is correct as only certain test cases can access entities. Make it stricter?